### PR TITLE
Pass in a mapping of the tools that a plugin has access to

### PR DESCRIPTION
**Pull request by abertelrud**
Pass in a mapping of the tools that a plugin has access to, which will work for both built tools and for archive-provided tools.

### Motivation:

Plugin scripts should have access to built as well as vended tools.

### Modifications:

- add a tool mapping to the TargetBuildContext and modify the lookup function to use it
- the function lookup a tool now returns a `Tool` structure instead of just a path; this is different from the proposal, and would need to be reconciled
- before invoking the plugin, construct a tool mapping for all the accessible tools (as determined by the plugin dependencies)
- modify the artifact filtering logic to only include the variants that match the passed-in target triple
- narrow the interface for artifact selection to include just the triple, and not the whole BuildParameters
- add more specific error messages from TargetBuildContext

### Result:

Build tool plugins now have access to the tools they have said they depend on.  This can in the future be extended to include other artifacts than executables.

The `Tool` structure doesn't yet contain the version field from the artifact archive, but that can be added in a later PR.  When we do that we should perhaps also make it comparable, so that the plugin can do different things depending on the tool version.  Although it's not clear how useful that will actually be, since the plugin and the tool artifact are updated together, so the plugin script would never see unexpected versions of the tool.  The version number is mostly for informational purposes.

rdar://74662891

### CODE REVIEW

**Comment by abertelrud** [comment]
@swift-ci please smoke test

File: Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift [hunk 1]
```diff
@@ -69,15 +69,32 @@ public final class TargetBuildContext: Decodable {
     /// target on which the package plugin target depends. Returns nil, but
     /// does not throw an error, if the tool isn't found. Plugins that re-
     /// quire the tool should emit an error diagnostic if it cannot be found.
-    public func lookupTool(named name: String) throws -> Path {
-        // TODO: Rather than just appending the name, this should instead use
-        // a mapping of tool names to paths (passed in from the context).
-        return self.toolsDir.appending(name)
+    public func tool(named name: String, line: UInt = #line) throws -> Tool {
+        if let tool = self.tools[name] { return tool }
```
Comment (tomerd) [inline Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift:73]: case sensitive okay?

File: Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift [hunk 2]
```diff
@@ -69,15 +69,32 @@ public final class TargetBuildContext: Decodable {
...
+public enum TargetBuildContextError: Error {
+    /// Could not find a tool with the given name. This could be either because
+    /// it doesn't exist, or because the plugin doesn't have a dependency on it.
+    case toolNotFound(name: String, line: UInt)
+}
+
+extension TargetBuildContextError: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .toolNotFound(let name, let line):
```
Comment (tomerd) [inline Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift:94]: nice touch with line#

File: Tests/FunctionalTests/PluginTests.swift [hunk 3]
```diff
@@ -107,4 +107,22 @@ class PluginTests: XCTestCase {
         }
         #endif
     }
+
+    func testUseOfVendedBinaryTool() throws {
+        // Check if the host compiler supports the '-entry-point-function-name' flag.  It's not needed for this test but is needed to build any executable from a package that uses tools version 999.0.
+        try XCTSkipUnless(doesHostSwiftCompilerSupportRenamingMainSymbol(), "skipping because host compiler doesn't support '-entry-point-function-name'")
```
Comment (tomerd) [inline Tests/FunctionalTests/PluginTests.swift:113]: intentional?

**Review by tomerd — APPROVED** [review]
very nice, couple of questions

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift:73]: We should discuss this.  I am pretty sure that the proposal said that it's case sensitive, but will double-check.

File: Tests/FunctionalTests/PluginTests.swift (hunk 3 above)
Comment (abertelrud) [reply Tests/FunctionalTests/PluginTests.swift:113]: Unfortunately yes.  Right now the 999.0 tools version enables the functionality of building executables using the new flag, as well as enabling plugins.  I'm wondering whether we should have a separate flag to enable it.  I can also change the tests to use libraries.

File: Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift:94]: I do still need to figure out if there's a way to install a custom handler for unhandled top-level errors, the way you can do with Objective-C exceptions.  That would clean up the stderr output too.

File: Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackagePlugin/PublicAPI/TargetBuildContext.swift:73]: The proposal doesn't say, but I think we should clarify it to be case sensitive.  Target and product names are case sensitive, so I think that would make sense.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Pull request merged**